python/ray/_private/net.py

```diff
@@ -0,0 +1,177 @@
+        # This is an undesirable workaround for some existing tests.
+        raise ValueError(
+            f"Address invalid or not reachable: {address}\n\n" +
+            repr(e)
+        )
+    return socket.socket(inet_address[0], kind)
+
+
+def _get_sock_stream_from_host(address):
+    """
+    Returns a socket.socket which can either be IPv4 or IPv6 with
+    socket.SOCK_STREAM.
+
+    Favor IPv4 but expand to IPv6 if IPv4 not available.
+    """
+    return _get_sock_from_host(address, socket.SOCK_STREAM)
+
+
+def _get_sock_dgram_from_host(address):
+    """
+    Returns a socket.socket which can either be IPv4 or IPv6 with
+    socket.SOCK_DGRAM.
+
+    Favor IPv4 but expand to IPv6 if IPv4 not available.
+    """
+    return _get_sock_from_host(address, socket.SOCK_DGRAM)
+
+
+def _get_private_ip_addresses():
+    """
+    Connects to Google DNS over IPv4, IPv6, or both to discover private IP
+    addresses.
+    """
+    private_ip_addresses = []
+    for public_host in ["2001:4860:4860::8888", "8.8.8.8"]:
+        if ":" in public_host and not socket.has_ipv6:
+            continue
+        try:
+            with closing(_get_sock_dgram_from_host(public_host)) as s:
+                # IPv4 may not be available so skip
+                if ":" not in public_host and not s.family == socket.AF_INET:
+                    continue
+                s.connect((public_host, 80))
+                # get the local iface IP used to connect against the public_host
+                private_ip_addresses.append(s.getsockname()[0])
+        except Exception as ignored:  # noqa: F841
+            pass
+    return private_ip_addresses
+
+
+def _get_socket_dualstack_fallback_single_stack_laddr(kind=socket.SOCK_STREAM):
+    """
+    Similar to socket.socket() except that if networking is dualstack it will
+    listen on all address family interfaces.
+
+    If single-stack, then the loopback interface is queried for which stack is
+    allowed (IPv4 or IPv6).
+    """
+    if socket.has_dualstack_ipv6():
+        sock = socket.socket(socket.AF_INET6, kind)
+        sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 0)
+        # set up to listen on all IPv4 and IPv6 interfaces with a single socket
+        return sock
+    else:
+        # fall back to supporting the host single network stack.
+        # (IPv4 or IPv6 only)
+        return _get_sock_from_host("localhost", kind)
+
+
+def _parse_ip_port(address):
+    """
+    Parses a str of ip:port and returns a List with (ip, port).
+
+    This uses str.rsplit(":", 1) to support DNS, IPv6, and IPv4 parsing.
+    """
+    # dashboard passes bytes instead of str so coerce
+    if isinstance(address, bytes):
+        address = address.decode('ascii')
+
+    if not isinstance(address, str):
+        raise ValueError(
+            "address type not str or bytes:\n\n" +
+            f"    type: {repr(type(address))}\n" +
+            f"   value: {repr(address)}"
+        )
```

Comment:
This should probably commented out because of the mock interfering during unit tests.